pyquil/quil.py

```diff
@@ -737,6 +737,18 @@ def __str__(self) -> str:
             )
         )
 
+    def __repr__(self) -> str:
+        string = f"shots: {self.num_shots}\n"
+        string += "program: \n   "
+        string += "\n   ".join(
+            itertools.chain(
+                (str(dg) for dg in self._defined_gates),
+                (str(instr) for instr in self.instructions),
+                [""],
+            )
+        )
```

Comment:
Just realized this little snippet also appears in `Program.out` (in addition to `__str__` as discussed out-of-band).

Maybe just the `itertools.chain` part could be moved into a private method like `_program_lines` (or maybe `_to_quil`), then each of `__repr__`, `__str__`, and `out` could call that and `join` on whatever is appropriate.

Or for that matter, I guess `Program.out` could just call `__str__` or vice versa since they appear to be identical.

Reply:
One other observation: this might be slightly annoying for really large programs with lots of lines to have it spew into the repl and/or ipython notebook every time you evaluate something that returns a Program. Some Python built-ins like `list` or `set` will truncate with a trailing ellipses past a certain number of elements. Maybe that could make sense here as well?  On the one hand, `__repr__` is meant to be an unambiguous representation of the object, but on the other hand dumping thousands of lines might not be that useful either.  On the other-other hand, 10,000 line quil programs are probably rare :-)

Reply:
The python documentation sounds like it can't decide whether `__repr__` should be used to aid a human in debugging, or for serialization.

Reply:
@appleby I don't think they're so rare. Even a 1000 line program is going to be annoying.